File: src/main/java/ru/ezhov/rocket/action/notification/PopupNotificationService.java
package ru.ezhov.rocket.action.notification;

import javax.swing.SwingUtilities;
import java.awt.Dimension;
import java.awt.Point;
import java.awt.Toolkit;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.util.LinkedList;

class PopupNotificationService implements NotificationService {
    private final LinkedList<Message> messages = new LinkedList<>();
    private static final int SPACE_BETWEEN_MESSAGES = 15;

    @Override
    public void show(NotificationType type, String text) {
        SwingUtilities.invokeLater(() -> createAndShowMessage(type, text));
    }

    private void createAndShowMessage(NotificationType type, String text) {
        final Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
        Dimension messageDimension = new Dimension(
                (int) (screenSize.width * 0.1),
                (int) (screenSize.height * 0.07)
        );
        Message message = new Message(type, 3000, text);
        Point point;
        if (messages.isEmpty()) {
            point = new Point(screenSize.width - SPACE_BETWEEN_MESSAGES - messageDimension.width,
                    screenSize.height - SPACE_BETWEEN_MESSAGES - messageDimension.height
            );
        } else {
            final Message messageLast = messages.getLast();
            point = new Point(screenSize.width - SPACE_BETWEEN_MESSAGES - messageDimension.width,
                    messageLast.getY() - SPACE_BETWEEN_MESSAGES - messageDimension.height
            );
        }

        message.addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosed(WindowEvent e) {
                messages.remove(e.getSource());
            }
        });
        messages.addLast(message);
        message.showMessage(point, messageDimension);
    }
}

File: src/main/kotlin/ru/ezhov/rocket/action/notification/PopupNotificationService.kt
package ru.ezhov.rocket.action.notification

import java.awt.Dimension
import java.awt.Point
import java.awt.Toolkit
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import java.util.*
import javax.swing.SwingUtilities

internal class PopupNotificationService : NotificationService {
    private val messages = LinkedList<Message>()
    override fun show(type: NotificationType, text: String) {
        SwingUtilities.invokeLater { createAndShowMessage(type, text) }
    }

    private fun createAndShowMessage(type: NotificationType, text: String) {
        val screenSize = Toolkit.getDefaultToolkit().screenSize
        val messageDimension = Dimension((screenSize.width * 0.1).toInt(), (screenSize.height * 0.07).toInt())
        val message = Message(type, 3000, text)
        val point = if (messages.isEmpty()) {
            Point(screenSize.width - SPACE_BETWEEN_MESSAGES - messageDimension.width,
                    screenSize.height - SPACE_BETWEEN_MESSAGES - messageDimension.height
            )
        } else {
            val messageLast = messages.last
            Point(screenSize.width - SPACE_BETWEEN_MESSAGES - messageDimension.width,
                    messageLast.y - SPACE_BETWEEN_MESSAGES - messageDimension.height
            )
        }
        message.addWindowListener(object : WindowAdapter() {
            override fun windowClosed(e: WindowEvent) {
                messages.remove(e.source)
            }
        })
        messages.addLast(message)
        message.showMessage(point, messageDimension)
    }

    companion object {
        private const val SPACE_BETWEEN_MESSAGES = 15
    }
}